evm/src/invariant_fuzz.rs

```diff
@@ -0,0 +1,360 @@
+}
+
+impl InvariantFuzzError {
+    fn new(
+        invariant_address: Address,
+        func: &Function,
+        abi: &Abi,
+        result: &bytes::Bytes,
+        inputs: &[(Address, (Address, Bytes))],
+    ) -> Self {
+        InvariantFuzzError {
+            test_error: proptest::test_runner::TestError::Fail(
+                format!(
+                    "{}, reason: '{}'",
+                    func.name,
+                    match foundry_utils::decode_revert(result.as_ref(), Some(abi)) {
+                        Ok(e) => e,
+                        Err(e) => e.to_string(),
+                    }
+                )
+                .into(),
+                inputs.to_vec(),
+            ),
+            return_reason: "".into(),
+            // return_reason: status,
+            revert_reason: foundry_utils::decode_revert(result.as_ref(), Some(abi))
+                .unwrap_or_default(),
+            addr: invariant_address,
+            func: func.short_signature().into(),
+        }
+    }
+}
+
+pub fn invariant_strat(
+    depth: usize,
+    senders: Vec<Address>,
+    contracts: BTreeMap<Address, (String, Abi)>,
+) -> BoxedStrategy<Vec<(Address, (Address, Bytes))>> {
+    let iters = 1..depth + 1;
+    proptest::collection::vec(gen_call(senders, contracts), iters).boxed()
+}
+
+fn gen_call(
+    senders: Vec<Address>,
+    contracts: BTreeMap<Address, (String, Abi)>,
+) -> BoxedStrategy<(Address, (Address, Bytes))> {
+    let random_contract = select_random_contract(contracts);
+    random_contract
+        .prop_flat_map(move |(contract, abi)| {
+            let func = select_random_function(abi);
+            let senders = senders.clone();
+            func.prop_flat_map(move |func| {
+                let sender = select_random_sender(senders.clone());
+                (sender, fuzz_calldata(contract, func))
+            })
+        })
+        .boxed()
+}
+
+fn select_random_sender(senders: Vec<Address>) -> impl Strategy<Value = Address> {
+    let selectors = any::<prop::sample::Selector>();
+    let senders_: Vec<Address> = senders.clone();
+
+    if !senders.is_empty() {
+        // todo should we do an union ? 80% selected 15% random + 0x0 address by default
+        selectors.prop_map(move |selector| *selector.select(&senders_)).boxed()
+    } else {
+        let fuzz = fuzz_param(&ParamType::Address);
+        fuzz.prop_map(move |selector| {
+            // assurance above
+            selector.into_address().unwrap()
+        })
+        .boxed()
+    }
+}
+
+fn select_random_contract(
+    contracts: BTreeMap<Address, (String, Abi)>,
+) -> impl Strategy<Value = (Address, Abi)> {
+    let selectors = any::<prop::sample::Selector>();
+    selectors.prop_map(move |selector| {
+        let res = selector.select(&contracts);
+        (*res.0, res.1 .1.clone())
+    })
+}
+
+fn select_random_function(abi: Abi) -> impl Strategy<Value = Function> {
+    let selectors = any::<prop::sample::Selector>();
```

Comment:
there is a non-zero chance there is an upstream bug here that will make it so we cant use this selector. iirc, it never selected the first function when there were only 2 functions in a contract. Not sure if it was an implementation bug or something weird with selector.

need to make a test to confirm